overlord/servicestate/quota_handlers.go

```diff
@@ -56,11 +56,11 @@ type QuotaControlAction struct {
 	// the "update" or the "create" actions.
 	AddSnaps []string `json:"snaps"`
 
-	// MemoryLimit is the memory limit for the quota group being controlled,
-	// either the initial limit the group is created with for the "create"
+	// ResourceLimits is the set of resource limits to set on the quota group.
+	// Either the initial limit the group is created with for the "create"
 	// action, or if non-zero for the "update" the memory limit, then the new
 	// value to be set.
-	MemoryLimit quantity.Size
+	ResourceLimits resources.QuotaResources `json:"resource-limits"`
```

Comment:
Changing this like this will break older snapd's that had on-going quota control changes when they were updated or reverted, this is because either new snapd will not have the right information to process the old task (since it would be missing `resource-limits`) and additionally, old snapd will not have the information it checks when it sees a new task (missing `MemoryLimit`). I think this is fine for this PR specifically because quota groups are still experimental and so if folks who have enabled them happen to miss a change/task to change them while also somehow updating/reverting snapd (which is unlikely but hypothetically possible) then it's much less fallout than if we had allowed them for everybody.

Although actually now come to look at it, we never actually defined key names for all these fields in json so we currently today have this not-great JSON serialization in the state of quota update tasks:

```json
{
  "id": "283496",
  "kind": "quota-control",
  "summary": "Update quota group \"foo\"",
  "status": 4,
  "data": {
    "quota-control-actions": [
      {
        "quota-name": "foo",
        "action": "update",
        "snaps": null,
        "MemoryLimit": 200000000,
        "ParentName": ""
      }
    ],
    "state-updated": {
      "boot-id": "6a8aa087-7b92-489f-b397-da695fc0ecb2"
    }
  },
  "change": "33255",
  "spawn-time": "2022-01-14T15:19:33.999798224-06:00",
  "ready-time": "2022-01-14T15:19:35.14102224-06:00",
  "doing-time": 1130588579
}
```

We should probably define `omitempty` for all of the fields in this struct and also make `ParentName` and `MemoryLimit` have proper lower-case key names too. Can be in this PR or a followup, but should be done before we make quota groups non-experimental.

Reply:
Great catch Ian! You are very correct about your observation and you highlight another problem. There are no json keys set for any values in QuotaResources either. I will change the keys and add omitempty to them.